connectors/migrations/20241212_backfill_github_data_source_folders.ts

```diff
@@ -0,0 +1,236 @@
+    const dataSourceConfig = dataSourceConfigFromConnector(connector);
+    const repositories = await GithubCodeRepository.findAll({
+      where: {
+        connectorId: connector.id,
+      },
+    });
+
+    for (const repository of repositories) {
+      // upsert repository as folder
+      // Throws if error
+      const repoFolderId = await upsertRepositoryFolderNode(
+        repository,
+        dataSourceConfig
+      );
+
+      // Upsert Code folder if we have some (file or directory)
+      const shouldCreateCodeFolder = await repositoryContainsCode(repository);
+      if (shouldCreateCodeFolder) {
+        await upsertCodeFolderNode(
+          repository.repoId,
+          repoFolderId,
+          dataSourceConfig
+        );
+      }
+
+      const directories = await GithubCodeDirectory.findAll({
+        where: {
+          connectorId: connector.id,
+        },
+      });
+      // Upsert directories as folders, in chunks
+      await concurrentExecutor(
+        directories,
+        async (directory) => {
+          await upsertDirectoryFolderNode(
+            directory,
+            repository.id,
+            dataSourceConfig
+          );
+        },
+        { concurrency: 16 }
+      );
+    }
+
+    // Upsert issues folder if we have some
+    const repoWithIssues = await getUniqueRepositoryIdsWithIssues(connector.id);
+    await concurrentExecutor(
+      repoWithIssues,
+      async (repoId) => {
+        // The node id is the same as the repoId
+        await upsertIssueFolderNode(
+          repoId,
+          getRepositoryInternalId(repoId),
+          dataSourceConfig
+        );
+      },
+      { concurrency: 16 }
+    );
+
+    // Upsert discussions folder if we have some
+    const repoWithDiscussions = await getUniqueRepositoryIdsWithDiscussions(
+      connector.id
+    );
+    await concurrentExecutor(
+      repoWithDiscussions,
+      async (repoId) => {
+        // The node id is the same as the repoId
+        await upsertDiscussionFolderNode(
+          repoId,
+          getRepositoryInternalId(repoId),
+          dataSourceConfig
+        );
+      },
+      { concurrency: 16 }
+    );
+  }
+}
+
+async function upsertRepositoryFolderNode(
+  repository: GithubCodeRepository,
+  dataSourceConfig: DataSourceConfig
+) {
+  const repoFolderId = repository.repoId;
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: getRepositoryInternalId(repository.repoId),
+    parents: [repoFolderId],
+    title: repository.repoName,
+    mimeType: getMimeTypeFromGithubContentNodeType("REPO_FULL"),
+  });
+  return repoFolderId;
+}
+
+async function upsertCodeFolderNode(
+  repositoryId: string,
+  repositoryNodeId: string,
+  dataSourceConfig: DataSourceConfig
+) {
+  const codeFolderId = getCodeRootInternalId(repositoryId);
+  await upsertDataSourceFolder({
+    dataSourceConfig,
+    folderId: codeFolderId,
+    parents: [codeFolderId, repositoryNodeId],
+    title: "Code",
```

Comment:
I saw what you answered in the [Slack thread](https://dust4ai.slack.com/archives/C050SM8NSPK/p1734518344104799), totally fine either way, I'm really not sure as to how this field will be used in the future 👍 
That could be a premonitory sign we need a `titleWithParentsContext` column @philipperolet